### DISCUSSION TITLE
g not persisting

### DISCUSSION THREAD

**Discussion by sjordahl** [pallets/flask#4458] [Q&A]
I've built a flask app following the official tutorial and am not seeing persistence with respect to the g object.  I'm attempting to persist a single Postgres connection in g.db, but the connection is being recreated on every request.  All roads that I've traveled in this regard suggest that I'm using the object per design.

Here's my db.py file:

```
import psycopg2
from flask import g

from os import environ as env


def get_db():
    print("g: ", g.__dict__)
    if 'db' not in g:
        if env['DB_CONFIG'] == 'local':
            print("Database: using standalone local database")
            g.db = psycopg2.connect(
                                user="postgres",
                                password="xxxxxxxx",
                                host="www.xxx.yyy.zzz",
                                port="5432",
                                database="ab")
            ...
    print("Database: returning g.db")
    print("g: ", g.__dict__)
    return g.db


def close_db(e=None):
    db = g.pop('db', None)

    if db is not None:
        db.close()


def init_app(app):
    app.teardown_appcontext(close_db)
```

As you'll notice, I've added some debug print statements to determine what's going on, and I've reduced the choices for DB_CONFIG as they are irrelevant.  This particular function is being called via Javascript's fetch function rather than through browser navigation (which shouldn't matter AFAIK).  When using this code, I get the following output each time I call the check view:

```
192.168.X.Y - - [18/Feb/2022 16:02:18] "POST /check HTTP/1.1" 200 -
g:  {}
Database: using standalone local database
Database: returning g.db
g:  {'db': <connection object at 0x7fcd4a583540; dsn: '
g:  {'db': <connection object at 0x7fd70d9e7540; dsn: 'user=postgres password=xxx dbname=ab host=www.xxx.yyy.zzz port=5432', closed: 0>}
192.168.X.Y - - [18/Feb/2022 16:02:22] "POST /check HTTP/1.1" 200 -
g:  {}
Database: using standalone local database
Database: returning g.db
g:  {'db': <connection object at 0x7fcd4a583540; dsn: '
g:  {'db': <connection object at 0x7fd70d9e7540; dsn: 'user=postgres password=xxx dbname=ab host=www.xxx.yyy.zzz port=5432', closed: 0>}
192.168.X.Y - - [18/Feb/2022 16:02:26] "POST /check HTTP/1.1" 200 -
g:  {}
Database: using standalone local database
Database: returning g.db
g:  {'db': <connection object at 0x7fcd4a583540; dsn: '
g:  {'db': <connection object at 0x7fd70d9e7540; dsn: 'user=postgres password=xxx dbname=ab host=www.xxx.yyy.zzz port=5432', closed: 0>}
192.168.X.Y - - [18/Feb/2022 16:02:29] "POST /check HTTP/1.1" 200 -
g:  {}
Database: using standalone local database
Database: returning g.db
g:  {'db': <connection object at 0x7fcd4a583540; dsn: '
g:  {'db': <connection object at 0x7fd70d9e7540; dsn: 'user=postgres password=xxx dbname=ab host=www.xxx.yyy.zzz port=5432', closed: 0>}
192.168.X.Y - - [18/Feb/2022 16:02:32] "POST /check HTTP/1.1" 200 -
g:  {}
Database: using standalone local database
Database: returning g.db
g:  {'db': <connection object at 0x7fcd4a583540; dsn: '
g:  {'db': <connection object at 0x7fd70d9e7540; dsn: 'user=postgres password=xxx dbname=ab host=www.xxx.yyy.zzz port=5432', closed: 0>}
192.168.X.Y - - [18/Feb/2022 16:02:35] "POST /check HTTP/1.1" 200 -
```

So according to this output it appears that g is not persisting.  I've thought about the potential thread locality of g, but I can run the code as many times as I want, greatly exceeding the number of workers, and still get the same output.

Any advice is greatly appreciated.

Flask ver: 2.0.3
Python ver: 3.9.7

**Comment by ThiefMaster** [comment]
`g` is request-scoped.

**Reply by sjordahl** [reply]
I was seeing that g was moved from request to app scope way back.  When did it move back to request?  And if that's the case, why does the current documentation suggest that it's app scoped?

**Reply by ThiefMaster** [reply]
It wasn't. The app context is also request-scoped, it can just exist without an active http request.

**Reply by davidism** [reply]
And it is intentional that a new connection is created for each request. Sharing global connections or reusing connections across requests will lead to all sorts of issues if not handled very carefully. If you use Flask-SQLAlchemy instead of the basic system shown in the tutorial, SQLAlchemy provides connection pooling so that connections are reused correctly. That said, until you are much, much, much bigger than the tutorial, you will not have any issues creating new connections each request.

**Reply by davidism** [reply]
If it is read-only and does not change, then you can just have a global. If it is data or state that will change, use something like [`multiprocessing.Manager`](https://docs.python.org/3/library/multiprocessing.html#managers) to control access to it.

For databases specifically, again, don't worry about connection reuse, or use SQLAlchemy.

**Reply by sjordahl** [reply]
OK.  Thanks for the clarification.  So there's really no reason to use g at all in this case then.  I should rewrite get_db to be the following, correct?:

```
def get_db():
        if env['DB_CONFIG'] == 'local':
            return psycopg2.connect(
                                user="postgres",
                                password="xxxxxxxx",
                                host="www.xxx.yyy.zzz",
                                port="5432",
                                database="ab")
            ...
```

**Reply by sjordahl** [reply]
Thank you for your assistance!  Very much appreciate you guys.

**Reply by davidism** [reply]
No, `g` is used to *cache* the connection *for the length of the request*. That way, you only create one connection *per request*.

**Reply by davidism** [reply]
This is all described in the tutorial, in fewer words: https://flask.palletsprojects.com/en/2.0.x/tutorial/database/#connect-to-the-database

**Reply by sjordahl** [reply]
Just looked at that link.  Thank you again!
